feat: image builder

mikejgray opened the pull request:
This pull request introduces a significant new feature: an image building pipeline for the Neon Hub, utilizing `debos` to create a custom Debian Bookworm-based image. The primary goal is to provide a user-friendly "first boot" experience, especially for users who may not be technically inclined.

Key additions and changes include:

- **Image Building with Debos:** Implements a `debos`-based build system (`hub-efi.yaml` and associated recipes) to generate a bootable amd64 UEFI image.
- **Simplified User Experience:** Focuses on a smooth initial setup, including features like automatic login for the default user (`neon`), pre-configured networking (`NetworkManager`), and automatic filesystem expansion.
- **Desktop Environment:** Includes the KDE Plasma desktop environment.
- **Core Functionality:** The image comes with essential services like SSH (OpenSSH server), display management (SDDM), and system utilities. The default user `neon` (password `neon`) has `sudo` privileges without a password.
- **Hardware Support:** Incorporates common firmware packages (iwlwifi, realtek, atheros) to improve hardware compatibility, particularly for networking.
- **Documentation:** Adds a `README.md` within the `debos` directory explaining the build process, prerequisites, image features, and instructions for testing with QEMU.
- **Iterative Development:** The pull request shows a history of numerous commits addressing fixes, refactoring, and enhancements to the user experience and build process.

This initiative aims to make Neon Hub more accessible by providing a ready-to-use operating system image with a guided initial setup.

mikejgray committed:
feat: hub image builder

mikejgray committed:
switch to debos for base image

mikejgray committed:
chore(docs): correction to the installation section

mikejgray committed:
feat: debos

mikejgray committed:
fix: restore scripts

mikejgray committed:
feat: debos recipe for installing neon-hub

mikejgray committed:
formatting

mikejgray committed:
shell compat

mikejgray committed:
fix

mikejgray committed:
fix

mikejgray committed:
fix

mikejgray committed:
fix

mikejgray committed:
refactor: bring ansible folder into overlays

mikejgray committed:
revert: refactor of ansible playbook

mikejgray committed:
fix

mikejgray committed:
fix

mikejgray committed:
fix

mikejgray committed:
fix

mikejgray committed:
fix

mikejgray committed:
fix

mikejgray committed:
more network config and audio

mikejgray commented:
Internet on my laptop was fixed by: `sudo apt install -y wpasupplicant firmware-iwlwifi firmware-realtek firmware-atheros net-tools iproute2`

Which is frustrating becauase I know almost all of that is already in the recipe...

mikejgray committed:
aim for a better ux

mikejgray committed:
fix

mikejgray committed:
correction

mikejgray committed:
expand filesystem automatically

mikejgray committed:
first boot user experience

mikejgray committed:
fix

NeonDaniel commented on debos/.gitignore:
```diff
@@ -0,0 +1,9 @@
+*.img
```
Any reason not to merge this with the root `.gitignore`?

mikejgray replied:
Nope, just an artifact of working in multiple places

NeonDaniel commented on debos/README.md:
````diff
@@ -0,0 +1,74 @@
+3.  Run `debos` using the provided YAML configuration file:
+
+    ```bash
+    sudo debos hub-efi.yaml
````
I *think* you can run this without `sudo` as long as the user is in the `kvm` group

NeonDaniel commented on debos/README.md:
```diff
@@ -0,0 +1,74 @@
+    *   Install and configure the GRUB UEFI bootloader.
+    *   Perform system configuration (user setup, SSH, services).
+
+4.  The resulting image file (`debian-uefi.img` by default) will be created in the current directory.
```
Consider changing this to `neon-hub-amd64_{timestamp}.img` or similar to more easily track multiple builds/output files. i.e. [neon-debos build](https://github.com/NeonGeckoCom/neon_debos/blob/eecaa7c4e8657aad0ab949878b28ea5945c0d738/run_docker_debos.sh#L64)

NeonDaniel commented on debos/overlays/auto-login/etc/systemd/system/serial-getty@.service.d/override.conf:
```diff
@@ -0,0 +1,3 @@
+[Service]
+ExecStart=
```
stray line to remove?

NeonDaniel commented on debos/qemu-boot-efi.sh:
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+qemu-system-x86_64 \
+    -enable-kvm -cpu host -smp 2 -m 1G \
```
Is there any benefit to increasing this to our minimum system requirements?

mikejgray replied:
I'm not sure I'll even keep this file, tbh, it just made testing easier early on

NeonDaniel reviewed (commented):
Some early feedback just based on reading through the changes

mikejgray committed:
fix first boot experience

mikejgray committed:
update to first boot experience

mikejgray committed:
not funny anymore

mikejgray committed:
address feedback

mikejgray committed:
remove outdated recipe code

mikejgray committed:
fix: issues found in testing

mikejgray committed:
getting real sick of this particular issue

mikejgray committed:
fixes for script after some testing

mikejgray committed:
splashscreen/wallpaper and put neon user in sudoers

mikejgray committed:
enhanced first boot experience

mikejgray committed:
auto-format error correction

mikejgray committed:
additional fixes

mikejgray committed:
remove unused scripts

mikejgray committed:
various ux improvements

mikejgray committed:
corrections

mikejgray committed:
more updates

mikejgray committed:
update debos readme

mikejgray committed:
more ux enhancements

mikejgray committed:
correct folder

mikejgray committed:
don't cause kde to not run

NeonDaniel commented:
Summary of testing feedback:
## Documentation
- Secure boot must be disabled
- Yacht default credentials should be in our docs somewhere (or at least a link to Yacht docs)
- Consider a README on the desktop with basic troubleshooting, like yacht login and a note that SSL errors for local services are expected

## Bugs
- "startkde: Could not start Plasma session." Error came up on first boot of a clean image for me twice. Resolution was to reboot. If there is no solution, then maybe just an FAQ note for now
- Prompted to continue setup upon every login (maybe related to no network connection until I entered a KDE Wallet password?)
- `/etc/rabbitmq/enabled_plugins` permission denied. Easily resolved with `chmod o+w /home/neon/xdg/config/rabbitmq` but better solution may be to figure out the UID running RMQ in that container
- After restarting `iris`, I can confirm audio is working but I am not getting any response and just an error when submitting audio input. For some reason, I can't get developer tools open in the browser to diagnose further
  - Same error for text input
- Nodes are unable to connect due to SSL errors. One possible solution would be to allow voice clients to connect without SSL, though I don't think this will work for Websockets

## Nit-Picking
- Notification to connect to the internet comes up before the desktop loads and is covered by the loading screen after being temporarily visible
- Consider generating an encryption key so KDE Wallet works ootb to save network credentials
- Consider some kind of ETA for how long installation will take, or else leaving some visual indicator that things are happening
- System time did not auto-update my timezone; consider a manual ntp sync and timezone update early in the installer
- Background service needed to be manually started, but after starting once the background came up on every subsequent boot
- Restarting containers via yacht seems to have not restarted anything, but restarting via CLI worked
- I could not get my laptop running the hub to resume from sleep. Ended up going for a hard reset (not an issue if this is just a power management issue for battery-powered devices)

mikejgray committed:
addressing feedback

mikejgray commented:
> Summary of testing feedback:
> 
> ## Documentation
> * Secure boot must be disabled
Noted in the README.
> * Yacht default credentials should be in our docs somewhere (or at least a link to Yacht docs)
It's in the docs, but I added it to another section just in case.
> * Consider a README on the desktop with basic troubleshooting, like yacht login and a note that SSL errors for local services are expected
Created a README.txt on the desktop.
> * > It looks like the documentation can just be updated to tell the user to run build.sh from within debos, or else update build.sh to cd to the script's parent directory
Updated per your suggestion.
> 
> ## Bugs
> * "startkde: Could not start Plasma session." Error came up on first boot of a clean image for me twice. Resolution was to reboot. If there is no solution, then maybe just an FAQ note for now
Noted in the README.
> * Prompted to continue setup upon every login (maybe related to no network connection until I entered a KDE Wallet password?)
Adjusted the boot script to look for a lock file before doing anything else, so that should resolve it. Please let me know if that's not the case.
> * `/etc/rabbitmq/enabled_plugins` permission denied. Easily resolved with `chmod o+w /home/neon/xdg/config/rabbitmq` but better solution may be to figure out the UID running RMQ in that container
UID:GID is 999:999. Added a step to the recipe to make sure those permissions are applied recursively to `/home/neon/xdg/config/rabbitmq`.
> * After restarting `iris`, I can confirm audio is working but I am not getting any response and just an error when submitting audio input. For some reason, I can't get developer tools open in the browser to diagnose further
>   
>   * Same error for text input
> * Nodes are unable to connect due to SSL errors. One possible solution would be to allow voice clients to connect without SSL, though I don't think this will work for Websockets
Opened a bug issue on the Iris repo, as audio recording will not be allowed in the browser if it's not using SSL, and this would impact the text input as well: https://github.com/NeonGeckoCom/neon-iris/issues/75
> 
> ## Nit-Picking
> * Notification to connect to the internet comes up before the desktop loads and is covered by the loading screen after being temporarily visible
I don't know a way around this. It's using the Freedesktop .desktop file spec and [I don't see any keys that would let us choose a loading order](https://specifications.freedesktop.org/desktop-entry-spec/latest/recognized-keys.html)
> * Consider generating an encryption key so KDE Wallet works ootb to save network credentials
Do you have docs on how to do this? I tried installing some packages I read would help and they did not.
> * Consider some kind of ETA for how long installation will take, or else leaving some visual indicator that things are happening
I'll look around for a KDE widget or some other way to check on Docker image status, but since the vast majority of the wait time is downloading large Docker containers, it's not easy to predict how long that will take. Disk type (USB vs SSD, for example) will influence it, as well internet connection speed.
> * System time did not auto-update my timezone; consider a manual ntp sync and timezone update early in the installer
Borrowed the update-clock section of the neon_debos repo.
> * Background service needed to be manually started, but after starting once the background came up on every subsequent boot
I'll mull this one over some more. I wish you could just set a config somewhere instead of having to run this dbus-dependent service to change the wallpaper, but here we are.
> * Restarting containers via yacht seems to have not restarted anything, but restarting via CLI worked
May want to evaluate a replacement for Yacht at some point...it unfortunately seems to be minimally maintained, which didn't appear to be the case when I chose it.
> * I could not get my laptop running the hub to resume from sleep. Ended up going for a hard reset (not an issue if this is just a power management issue for battery-powered devices)
Open to suggestions on how to deal with this - my assumption is a Hub will not be running on battery power most of the time, so it may be a non-issue.

NeonDaniel commented on debos/overlays/neon-hub-desktop-files/home/neon/Desktop/README.txt:
```diff
@@ -0,0 +1,20 @@
+Libretranslate Translation Service: https://libretranslate.neon-hub.local
+RabbitMQ Administration Page: https://rmq-admin.neon-hub.local
+
+The two IRIS services require additional setup. See the official documentation for instructions for more technical users: hhttps://neongeckocom.github.io/neon-hub-installer/service-details/#custom-ssl-certificate
```
Minor typo
```suggestion
The two IRIS services require additional setup. See the official documentation for instructions for more technical users: https://neongeckocom.github.io/neon-hub-installer/service-details/#custom-ssl-certificate
```

mikejgray committed:
Update debos/overlays/neon-hub-desktop-files/home/neon/Desktop/README.txt

Co-authored-by: Daniel McKnight <34697904+NeonDaniel@users.noreply.github.com>

NeonDaniel commented:
> Adjusted the boot script to look for a lock file before doing anything else, so that should resolve it. Please let me know if that's not the case.

Looks to be fixed for me too

> UID:GID is 999:999. Added a step to the recipe to make sure those permissions are applied recursively to /home/neon/xdg/config/rabbitmq.

I'm still seeing that directory (and its contents) owned by `neon:neon` (UID 1000)

> Do you have docs on how to do this? I tried installing some packages I read would help and they did not.

Not off the top of my head, but I'll do a little reading up on this one

> Borrowed the update-clock section of the neon_debos repo.

I forgot that also isn't handling TZ detection, I just have it set to Pacific Time so I never thought about it. Should be fine for now; the time is correct so SSL, etc. should work

> I'll mull this one over some more. I wish you could just set a config somewhere instead of having to run this dbus-dependent service to change the wallpaper, but here we are.

Claude or GPT might be able to help pick a dependent service to start this after.. It seems like `After="graphical.target"` should be sufficient, but I know that isn't always the case.

> Open to suggestions on how to deal with this - my assumption is a Hub will not be running on battery power most of the time, so it may be a non-issue.

Yeah, not a priority at this point as long as powered devices stay awake

> May want to evaluate a replacement for Yacht at some point...it unfortunately seems to be minimally maintained, which didn't appear to be the case when I chose it.

[dockge](https://github.com/louislam/dockge) is another option I know of.. It looks like there's some activity there too, though the most recent commits are mostly auto-translation of the UI

mikejgray committed:
address feedback

mikejgray commented:
> > UID:GID is 999:999. Added a step to the recipe to make sure those permissions are applied recursively to /home/neon/xdg/config/rabbitmq.
> 
> I'm still seeing that directory (and its contents) owned by `neon:neon` (UID 1000)

After you shared this, I realized I put the chown in the wrong recipe. It should be fixed now.

> > I'll mull this one over some more. I wish you could just set a config somewhere instead of having to run this dbus-dependent service to change the wallpaper, but here we are.
> 
> Claude or GPT might be able to help pick a dependent service to start this after.. It seems like `After="graphical.target"` should be sufficient, but I know that isn't always the case.

I changed my approach to use a set-default-wallpaper.sh file in /usr/local/bin, and have a .desktop file run that once. The script deletes the .desktop file afterwards but the script remains. Worked well in testing on a local KDE Debian VM.

> > May want to evaluate a replacement for Yacht at some point...it unfortunately seems to be minimally maintained, which didn't appear to be the case when I chose it.
> 
> [dockge](https://github.com/louislam/dockge) is another option I know of.. It looks like there's some activity there too, though the most recent commits are mostly auto-translation of the UI

There was some reason I didn't like dockge, but the reason is escaping me...might be worth a second look

mikejgray committed:
address startkde error

mikejgray committed:
fix busted sddm.service

mikejgray committed:
chore: use the neon icon for desktop links

NeonDaniel commented on debos/overlays/ansible/templates/nginx.conf.j2:
One small change to fix WS connections. In the `hana.neon-hub.local` location block, we need:
```
            proxy_set_header Connection "upgrade";
            proxy_set_header Upgrade $http_upgrade;
```

mikejgray committed:
node fixes

NeonDaniel reviewed (approved):
Tested a clean image and confirmed both HTTP and WS clients work as expected
